c/src/lib/libbsp/arm/shared/stm32fxxxx/uart/hal-uart.c

```diff
@@ -0,0 +1,647 @@
+
+    switch ( uartType ) {
+
+    case STM32F_UART_MODE_POLLING:
+      ret = HAL_UART_Receive(hUartHandle, buf, (uint16_t) len, timeout);
+      break;
+
+    case STM32F_UART_MODE_INT:
+      ret = HAL_UART_Receive_IT(hUartHandle, buf, (uint16_t) len);
+      break;
+
+    case STM32F_UART_MODE_DMA:
+      ret = HAL_UART_Receive_DMA(hUartHandle, buf, (uint16_t) len);
+      break;
+    }
+  }
+
+  return ret;
+}
+
+static rtems_task stm32_uart_tx_task(
+  rtems_task_argument arg
+)
+{
+  size_t len;
+  rtems_status_code sc;
+  rtems_event_set   events_rx;
+  stm32_uart_driver_entry * pUart = (stm32_uart_driver_entry *) arg;
+  rtems_interval ticks_per_ms;
+
+  static uint32_t failure_counter = 0UL, success_counter = 0UL;
+
+  // Calculate the ticks per ms.  This is constant during execution
+  // of the application and therefore can be calculated just once
+  // at task startup.
+  ticks_per_ms = rtems_clock_get_ticks_per_second() / 1000;
+
+  while ( 1 ) {
+
+    // wait until there is a TX request
+    sc = rtems_message_queue_receive(pUart->tx_msg_queue,
+      (void*) &tx_msg[pUart->instance],
+      &len,
+      RTEMS_WAIT,
+      RTEMS_NO_TIMEOUT);
+    _Assert(len <= sizeof(tx_msg));
+
+    // send tx data to UART hardware
+    if ( sc == RTEMS_SUCCESSFUL ) {
+      (void) stm32_uart_transmit(pUart->base_driver_info.handle,
+        pUart->base_driver_info.uart_mode, (uint8_t*) tx_msg[pUart->instance],
+        len);
+    }
+
+    // Wait for event from TX complete callback before
+    // processing any additional messages.
+    sc = rtems_event_receive(UART_TX_DONE,
+                             RTEMS_WAIT | RTEMS_EVENT_ANY,
+                             ticks_per_ms * STM32F_MAX_UART_TX_TIME_ms,
+                             &events_rx);
+
+    if (sc != RTEMS_SUCCESSFUL) {
+      failure_counter++;
+      //stm32f_error_handler();
+    } else {
+      success_counter++;
+    }
+  }
+}
+
+static void uart_obtain(
+  stm32_uart_driver_entry *pUart
+)
+{
+  rtems_status_code sc;
+
+  sc = rtems_semaphore_obtain(pUart->mutex, RTEMS_WAIT, RTEMS_NO_TIMEOUT);
+  _Assert(sc == RTEMS_SUCCESSFUL);
+  (void) sc;
+}
+
+static void uart_release(
+  stm32_uart_driver_entry * pUart
+)
+{
+  rtems_status_code sc;
+
+  sc = rtems_semaphore_release(pUart->mutex);
+  _Assert(sc == RTEMS_SUCCESSFUL);
+  (void) sc;
+}
+
+static ssize_t stm32_uart_read(
+  rtems_libio_t *iop,
+  void *buffer,
+  size_t count
+)
+{
+  stm32_uart_device *pUartDevice = (stm32_uart_device *)IMFS_generic_get_context_by_iop(iop);
+  ssize_t ret_size = 0;
+  HAL_StatusTypeDef ret;
+  rtems_status_code sc;
+  rtems_event_set  events_rx;
+
+
+  if ( count <= MAX_UART_RX_MSG_SIZE ) {
+
+    // collect the requested number of characters from the UART
+    ret = stm32_uart_receive(pUartDevice->pUart->base_driver_info.handle,
+      pUartDevice->pUart->base_driver_info.uart_mode,
+      (uint8_t*) &(rx_msg[pUartDevice->pUart->instance]),
```

Comment:
I think this is silly to use the locally declared RX buffers -- I think I should use the supplied void* buffer passed into the function (and remove the subsequent memcpy)